numpy/core/overrides.py

```diff
@@ -13,8 +13,12 @@
     int(os.environ.get('NUMPY_EXPERIMENTAL_ARRAY_FUNCTION', 1)))
 
 
-add_docstring(
-    implement_array_function,
+# We can't use add_newdoc (which is more robust than add_docstring) here to
+# attach these docstrings (cyclic imports), and add_docstring() gives a
```

Comment:
Can we resolve these cyclic imports instead?

Reply:
I tried for 15 minutes, it's nontrivial. The other option is to move the docstring into the C code. I don't think `add_docstring` should be used ever.

I'm going for a minimal fix for 1.19.x though, this is a painful issue.

Reply:
> Or alternatively, just move these docstrings to add_newdocs with all the others?

Tried that too, doesn't work - `add_newdoc` cannot attached things to `_multiarray_umath`.

Reply:
> Tried that too, doesn't work - `add_newdoc` cannot attached things to `_multiarray_umath`.

I don't know what you mean by that - isn't that exactly what it does already?

Reply:
Question: This is about reloading and rerunning the same code twice. How about fixing this in `add_docstring` itself...

```
        doc_attr = PyObject_GetAttrString(obj, "__doc__");
        if (doc_attr != NULL && doc_attr != Py_None) {
            PyErr_Format(PyExc_RuntimeError, "object %s", msg);
            return NULL;
        }
```
replace that with a check that the old and new docstring are identical. And if they are simply do nothing.

Reply:
We won't notice if we set it twice by accident, but since it will still error if there is a disagreement, that seems not a huge issue and feels more reasonable then a try/except here that may hide errors.

Reply:
I still don't really follow why this file does its own docstring attachment - can't we do it just like we do here:

https://github.com/numpy/numpy/blob/84a4c4b03a43c915151193685588f80759f2ea44/numpy/core/_add_newdocs.py#L5846

In both cases, it's attaching a docstring to an extension method from multiarray_umath.

Reply:
> Question: This is about reloading and rerunning the same code twice. How about fixing this in `add_docstring` itself...
> 
> ```
>         doc_attr = PyObject_GetAttrString(obj, "__doc__");
>         if (doc_attr != NULL && doc_attr != Py_None) {
>             PyErr_Format(PyExc_RuntimeError, "object %s", msg);
>             return NULL;
>         }
> ```
> 
> replace that with a check that the old and new docstring are identical. And if they are simply do nothing.

yes this would be really nice. another quick and hacky option maybe just filter "already has a docstring" exceptions and raise everything else ? this will avoid hiding other exceptions, but still mask errors with same msg caused by something else.